### PR TITLE
feat: Add support for Azure OpenAI Service

### PR DESCRIPTION
**Pull request by tao12345666333**
fixes: #95

### CODE REVIEW

File: autochain/models/chat_openai.py [hunk 1]
```diff
@@ -230,3 +230,129 @@ def _create_llm_result(self, response: Mapping[str, Any]) -> LLMResult:
         llm_output = {"token_usage": response["usage"], "model_name": self.model_name}
         result = LLMResult(generations=generations, llm_output=llm_output)
         return result
+
+
+class ChatAzureOpenAI(BaseLanguageModel):
```
Comment (yyiilluu) [inline autochain/models/chat_openai.py:235]: let's combine this into the same ChatOpenAI model. 
we would need to add the following to the model 
if api type is azure, validate all other information existed set them to openai at `validate_environment` step. 
```
api_type: Optional[str] = None
azure_api_base: Optional[str] = None
azure_api_verision : Optional[str] = None
azure_deployment_name: : Optional[str] = None
```

File: autochain/models/chat_openai.py [hunk 2]
```diff
@@ -230,3 +230,129 @@ def _create_llm_result(self, response: Mapping[str, Any]) -> LLMResult:
...
+        response = self.generate_with_retry(**generation_param)
+        return self._create_llm_result(response)
+
+    def _create_message_dicts(
+        self,
+        messages: List[BaseMessage],
+        tools: Optional[List[Tool]],
+        stop: Optional[List[str]],
+    ) -> Tuple[List[Dict[str, Any]], List[Dict[str, Any]], Dict[str, Any]]:
+        params: Dict[str, Any] = {**{"model": self.model_name, "engine": self.deployment_name}, **self._default_params}
```
Comment (yyiilluu) [inline autochain/models/chat_openai.py:339]: to reuse the same model, update `params` if self.azure_deployment_name is not None and self.api_type == "azure"

**Review by yyiilluu — CHANGES_REQUESTED** [review]

File: autochain/models/chat_openai.py (hunk 1 above)
Comment (tao12345666333) [reply autochain/models/chat_openai.py:235]: Thanks for your suggestions.

I have a point to discuss.

We know that both `api_version` and 
`deployment_name` exist for Azure, so adding `azure_` prefix is ​​normal. 

But for `api_base`, if the user uses a proxy to access OpenAI, the value can also be modified.

I think we can use `api_base` like `api_type` or use `openai_api_base`

WDYT？

Comment (yyiilluu) [reply autochain/models/chat_openai.py:235]: makes sense. please put some comments for each parameter

**Comment by yyiilluu** [comment]
thanks!

**Pull request merged**